### PR TITLE
Zephyr support

### PR DESCRIPTION
**Pull request by mlaz**
This PR aims to grant Zephyr mcumgr basic functionality.

### CODE REVIEW

**Comment by mlaz** [comment]
Depends on https://github.com/zephyrproject-rtos/zephyr/pull/20584

**Comment by nvlsianpu** [comment]
I saw in the code `that img_mgmt_impl_erase_if_needed()` is used for lazy erase, but couldn't found place where erase of image status is ensured (it land in the last sector (or sectors) of the image slot).

**Comment by nvlsianpu** [comment]
^^ @mlaz In zephy this is ensured here:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/dfu/img_util/flash_img.c#L205-L213

**Comment by mlaz** [comment]
@nvlsianpu I believe the image status is set here: https://github.com/apache/mynewt-mcumgr/blob/cbbbb45d81aef966dc417272904fe7a6a9e321f3/cmd/img_mgmt/src/img_mgmt.c#L503

**Comment by nvlsianpu** [comment]
@mlaz 
I believe you mix `image trailer` with `image header + image metadata`
Signing operation adds to the executable binary the image header with required info abut image in TLV format, both glued to the executable payload.

Image trailer is something else - usually it is not included in the update image as it lands in the last flash page of a image slot (or a few very last pages if pages are small).

Looks like `img_mgmt_dfu_pending();` is able to set pending status on a image which is operation of writing something to the image trailer flasgs.

**Comment by mlaz** [comment]
Taking into account the review changes were made so it now also depends on https://github.com/zephyrproject-rtos/zephyr/pull/20973

**Comment by de-nordic** [comment]
When this will be merged?

**Comment by mlaz** [comment]
@nvlsianpu ^

**Comment by nvlsianpu** [comment]
I had tested this extensivelly with https://github.com/zephyrproject-rtos/zephyr/pull/20584
Looks like BLE transport works well.
Unfortunately shell transport has regression - I was able to finish upload only one time in few trial while with zephyr master version it works just perfectly. @de-nordic can you help to find regression reason.

File: cborattr/include/cborattr/cborattr.h [hunk 1]
```diff
@@ -26,7 +26,12 @@
 #include <ctype.h>
 #include <stdio.h>
 #include <sys/types.h>
+
+#ifdef __ZEPHYR__
+#include <cbor.h>
+#else
 #include "tinycbor/cbor.h"
+#endif
```
Comment (vrahane) [inline cborattr/include/cborattr/cborattr.h]: In the zephyr fork of tinycbor, they have made changes to include header files the mynewt way. Here is the commit which addresses that:
https://github.com/zephyrproject-rtos/tinycbor/commit/0fc68fceacd1efc1ce809c5880c380f3d98b7b6e
We should just keep what was there.

File: cborattr/src/cborattr.c [hunk 2]
```diff
@@ -17,24 +17,25 @@
...
 #include "cborattr/cborattr.h"
-#include "tinycbor/cbor.h"
-#include "tinycbor/cbor_buf_reader.h"
+
+#ifdef __ZEPHYR__
+#include <zephyr.h>
+#include <cbor.h>
+#include <cbor_buf_reader.h>
+#define CBORATTR_MAX_SIZE CONFIG_MGMT_CBORATTR_MAX_SIZE
+#endif
```
Comment (vrahane) [inline cborattr/src/cborattr.c]: Same comment as above, the constant can stay and so would the zephyr include.

File: cmd/img_mgmt/src/img_mgmt_state.c [hunk 3]
```diff
@@ -18,7 +18,12 @@
  */
 
 #include <assert.h>
+#ifdef __ZEPHYR__
+#include <cbor.h>
+#else
 #include "tinycbor/cbor.h"
+#endif
```
Comment (vrahane) [inline cmd/img_mgmt/src/img_mgmt_state.c]: Same comment as above

File: cmd/os_mgmt/src/os_mgmt.c [hunk 4]
```diff
@@ -19,7 +19,13 @@
 
 #include <assert.h>
 #include <string.h>
+
+#ifdef __ZEPHYR__
+#include <cbor.h>
+#else
 #include "tinycbor/cbor.h"
+#endif
```
Comment (vrahane) [inline cmd/os_mgmt/src/os_mgmt.c]: Same comment as above

File: mgmt/include/mgmt/mgmt.h [hunk 5]
```diff
@@ -21,7 +21,12 @@
 #define H_MGMT_MGMT_
 
 #include <inttypes.h>
+
+#ifdef __ZEPHYR__
+#include <cbor.h>
+#else
 #include "tinycbor/cbor.h"
+#endif
```
Comment (vrahane) [inline mgmt/include/mgmt/mgmt.h]: Same comment as above

File: mgmt/src/mgmt.c [hunk 6]
```diff
@@ -18,7 +18,13 @@
  */
 
 #include <string.h>
+
+#ifdef __ZEPHYR__
+#include <cbor.h>
+#else
 #include "tinycbor/cbor.h"
+#endif
+
```
Comment (vrahane) [inline mgmt/src/mgmt.c]: Same comment as above

File: smp/src/smp.c [hunk 7]
```diff
@@ -22,10 +22,15 @@
 #include <assert.h>
 #include <string.h>
 
+#ifdef __ZEPHYR__
+#include <cbor.h>
+#else
+#include "tinycbor/cbor.h"
+#endif
```
Comment (vrahane) [inline smp/src/smp.c]: Same comment as above

**Comment by utzig** [comment]
> Unfortunately shell transport has regression - I was able to finish upload only one time in few trial while with zephyr master version it works just perfectly. @de-nordic can you help to find regression reason.

Are you looking into this @de-nordic ?

**Comment by mlaz** [comment]
> Unfortunately shell transport has regression - I was able to finish upload only one time in few trial while with zephyr master version it works just perfectly. @de-nordic can you help to find regression reason.

@nvlsianpu I have been trying to reproduce that bug and I can upload every time via serial, both erasing progressively and doing a single step erase. I am using a serial adapter, connected to the serial pins on an nRF52dk. Once I try using the J-Link serial line the behaviour gets unstable using most of the mcumgr commands. From this output snippet you may see that probably the shell running on the uart is causing the problem, I don't think this is fixable on the mcumgr side.
```
DEBU[2019-12-10 17:44:44.861] Tx serial
00000000  41 41 41 41 41 41 41 41  41 41 41 41 41 41 41 41  |AAAAAAAAAAAAAAAA|
00000010  41 41 41 41 41 41 41 41  41 41 41 41 41 41 41 41  |AAAAAAAAAAAAAAAA|
00000020  41 41 41 41 41 41 41 41  41 41 41 41 41 41 41 41  |AAAAAAAAAAAAAAAA|
00000030  41 47 56 70 62 57 46 6e  5a 51 42 6a 62 47 56 75  |AGVpbWFnZQBjbGVu|
00000040  47 53 6c 67 59 32 39 6d  5a 67 42 6a 63 32 68 68  |GSlgY29mZgBjc2hh|
00000050  57 43 43 6d 35 66 65 71  50 41 58 6a 4e 71 38 79  |WCCm5feqPAXjNq8y|
00000060  4a 44 64 58 7a 72 69 6d  35 66 64 6b 37 79 71 4a  |JDdXzrim5fdk7yqJ|
00000070  32 4e 46 64 69 42 57 4f  49 38 2f 48              |2NFdiBWOI8/H| 
DEBU[2019-12-10 17:44:44.861] Tx serial
00000000  0a                                                |.| 
DEBU[2019-12-10 17:44:44.867] Rx serial:
00000000  1b 5b 6d 1b 5b 31 3b 33  32 6d 75 61 72 74 3a 7e  |.[m.[1;32muart:~|
00000010  24 20 1b 5b 6d 55 53 42  53                       |$ .[mUSBS| 
DEBU[2019-12-10 17:44:44.87] Rx serial:
00000000  1b 5b 31 3b 33 31 6d 55  53 42 53 3a 20 63 6f 6d  |.[1;31mUSBS: com|
00000010  6d 61 6e 64 20 6e 6f 74  20 66 6f 75 6e 64        |mand not found| 
DEBU[2019-12-10 17:44:44.873] Rx serial:
```

**Comment by de-nordic** [comment]
> Are you looking into this @de-nordic ?

I am looking.

**Comment by de-nordic** [comment]
OK. So I have tested following branches mcumgr_changes@zephyr  and run_zephyr@modules/lib/mcumgr. No rebasing done.

SMP over shell, but no terminal connected to the shell, from  PC, at the same time when MCUMGR is running on the PC.

Using script:
`$$ UART=/dev/ttyACM0; BAUD=57600; while [ ! -e ~/tmp/mcumgr_stop ]; do time sudo ~/go/bin/mcumgr -t 1200 --conntype serial --connstring dev=$UART,baud=$BAUD image upload ~/nordic-src/ncs/nrf52840_pca10056/Disco.bin; if [ $? -ne 0 ]; then echo "Error"; fi;  sudo ~/go/bin/mcumgr -t 1200 --conntype serial --connstring dev=$UART,baud=$BAUD image list; done`

I have counted errors or total stoppages. 

In case when using board USB serial (/dev/ttyACM0) I have faced some timeouts on binary every few transfers when baud=115200; after modifying DTS file to use 57600 speed (and MCUMGR)  have 100% success ratio in transfers.

When I have connected directly to Tx/Rx of IC, bypassing the USB interface on board, both baud rates have me 100% success ratio in transmission.

I have not seen regression in transfer success rate.

**Comment by mlaz** [comment]
So far we've only detected malfunction when testing using the usb port(J-Link uses it, it creates an ttyACM device using uart0) on an nRF52dk/nRF52840dk opposed to using a serial adaptor on the configured pins. This happens both on Zephyr and Mynewt. This is being looked into at the moment, @de-nordic mentioned that probably it would be needed do enable hw flow control, however mcumr-cli currently doesn't support this(the golang serial lib we are using doesn't support flow control).

**Comment by de-nordic** [comment]
> 
> 
> So far we've only detected malfunction when testing using the usb port(J-Link uses it, it creates an ttyACM device using uart0) on an nRF52dk/nRF52840dk opposed to using a serial adaptor on the configured pins. This happens both on Zephyr and Mynewt. This is being looked into at the moment, @de-nordic mentioned that probably it would be needed do enable hw flow control, however mcumr-cli currently doesn't support this(the golang serial lib we are using doesn't support flow control).

I have been working with @mlaz on this, mostly testing his solution. I have done tests on nRF52840 and found out that SMP via ttyACM works unreliably while direct connection to IC with use of FTDI dongle is OK. 
In case of Miguel's tests, his board (correct me here @mlaz) has not been able to successfully perform any transfers via ttyACM while  direct connection via FTDI worked.

File: cborattr/src/cborattr.c [hunk 8]
```diff
@@ -17,24 +17,29 @@
...
  */
 
-//#include <syscfg/syscfg.h>
 #include "cborattr/cborattr.h"
 #include "tinycbor/cbor.h"
 #include "tinycbor/cbor_buf_reader.h"
 
+#ifdef __ZEPHYR__
+#include <zephyr.h>
+<<<<<<< Updated upstream
```
Comment (de-nordic) [inline cborattr/src/cborattr.c]: I think you left this by accident.

Comment (mlaz) [reply cborattr/src/cborattr.c]: you are right

Comment (mlaz) [reply cborattr/src/cborattr.c]: I just checked my code locally, that is not here also the comment is marked as outdated.

**Comment by utzig** [comment]
I decided to try this as well and had to apply a few local changes to build:

to cborattr/src/cborattr.c

```
-#define CBORATTR_MAX_SIZE CONFIG_MGMT_CBORATTR_MAX_SIZE
+#define CBORATTR_MAX_SIZE 512
```
because the Kconfig with the option was removed.

cmd/img_mgmt/src/img_mgmt.c

```
-    return img_mgmt_read_info(boot_current_slot, ver, NULL, NULL);
+    return img_mgmt_read_info(0, ver, NULL, NULL);
```

I don't know why it did not find this MCUBoot symbol, but since this is the running image and MCUBoot at least for the time being only supports images running in slot0 it is safe to use `0` there.

I was able to reproduce the failures reported using the nrf52840_pca10056. No idea why because looking at the diff there is nothing that looks like it would break the serial communication; I will test with current mcumgr on Zephyr as well; I would expect it to have those same fails.

**Comment by utzig** [comment]
Ok, so I tested now with current mcumgr (https://github.com/zephyrproject-rtos/mcumgr) and had no fails at all after 10 or so uploads. The main difference I see when uploading an image is that with the current mcumgr it takes a long time to start (something like 10s maybe), while this repo/PR starts almost instantaneously. Maybe one is erasing everything first and the other is doing an erase+write loop?

**Comment by de-nordic** [comment]
> 
> 
> I decided to try this as well and had to apply a few local changes to build:
> 
> (...)
> 
> I don't know why it did not find this MCUBoot symbol, but since this is the running image and MCUBoot at least for the time being only supports images running in slot0 it is safe to use `0` there.
> 
> 
I think that the PR is missing information that it requires parallel changes to zephyr that are reviewed here:
https://github.com/zephyrproject-rtos/zephyr/pull/20584

**Comment by de-nordic** [comment]
Yesterday I have done some more changes and can confirm @utzig findings regarding https://github.com/zephyrproject-rtos/mcumgr (vanilla manifest-rev branch), though I left it alone for some time and did more than 100 successful uploads with no failures.

I have run the solution from this PR and got at most 16 uploads, with progressive erase note enabled until it failed.

NOTE: Prior to testing it is good disable storage driver for JLink to make it more stable in behaviour: https://docs.zephyrproject.org/latest/guides/tools/nordic_segger.html?highlight=MSDDisable#disabling-the-mass-storage-device-functionality, if you have not done it before.

**Comment by utzig** [comment]
> I think that the PR is missing information that it requires parallel changes to zephyr that are reviewed here:
zephyrproject-rtos/zephyr#20584

Thanks, with that PR the build issues are solved. I also tested this mcumgr PR with Mynewt and it is flawless (no fails), so I would assume there is some Zephyr specific change causing the timeouts.

**Comment by de-nordic** [comment]
@utzig thanks for getting involved with testing it with on Mynewt and Zephyr.

**Comment by mlaz** [comment]
@utzig @de-nordic Thanks for the testing/insights. I believe that now the related PRs may be merged, right @vrahane @nvlsianpu ?

**Comment by utzig** [comment]
Another bit of trivia which might be useful: disabling BLE transport (`CONFIG_MCUMGR_SMP_BT=n`) does stop timeouts from happening (at least in my tests).

**Comment by vrahane** [comment]
@utzig and @nvlsianpu Thanks for the extensive testing and review. @utzig The timeouts issue that goes away on disabling the BLE transport, do you also see the timeouts in earlier version of mcumgr ?

**Comment by utzig** [comment]
> @utzig The timeouts issue that goes away on disabling the BLE transport, do you also see the timeouts in earlier version of mcumgr?

The timeouts don't happen using https://github.com/zephyrproject-rtos/mcumgr

**Comment by de-nordic** [comment]
I have rerun some tests, taking into account https://github.com/apache/mynewt-mcumgr/pull/46#issuecomment-567141118, and done some non-shell SMP testing and BLE check. For me it seems fine.

**Comment by utzig** [comment]
> I have rerun some tests, taking into account #46 (comment), and done some non-shell SMP testing and BLE check. For me it seems fine.

@nvlsianpu @vrahane Since we know the circumstances under which the timeouts happen, would it be OK to merge this PR as is, and then investigate/fix in a further PR?

**Comment by vrahane** [comment]
I am fine with merging this PR, I think the timeouts can be fixed later in the zephyr port. Do you agree @nvlsianpu ?

**Pull request merged**